### PR TITLE
yumeko market schema

### PR DESCRIPTION
**Pull request by horlami228**

### CODE REVIEW

**Comment by horlami228** [comment]
Screenshot from 2024-11-26 18-56-50
Screenshot from 2024-11-26 18-56-39
Screenshot from 2024-11-26 18-57-50

File: modules/market/market.models.ts [hunk 1]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  source: { type: String, required: true },
+  summary: { type: String, required: true },
+  labels: {
+    sector: { type: String },
+    topic: { type: String },
+    sentiment: { type: String },
+  },
+});
+
+export const InvestorProfile = mongo.createModel<Types.InvestorProfileDocument>('InvestorProfile', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: change this to MarketInvestor

File: modules/market/market.models.ts [hunk 2]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  name: { type: String, required: true },
+  portfolios: { type: mongo.Schema.Types.ObjectId, required: true },
+  totalPnL: { type: Number },
+});
+
+export const InvestmentPortfolio = mongo.createModel<Types.InvestmentPortfolioDocument>('InvestmentPortfolio', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: MarketInvestmentPortfolio

File: modules/market/market.models.ts [hunk 3]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+      goalPercentage: { type: Number, required: true },
+      currentPnL: { type: Number },
+      historicalPnL: { type: Number },
+    },
+  ],
+  investments: { type: mongo.Schema.Types.ObjectId, required: true },
+  totalPnL: { type: Number },
+});
+
+export const Investment = mongo.createModel<Types.InvestmentDocument>('Investment', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: MarketInvestment

File: modules/market/market.models.ts [hunk 4]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+      type: { type: String, required: true },
+      price: { type: Number, required: true },
+      quantity: { type: Number, required: true },
+    },
+  ],
+  stock: { type: mongo.Schema.Types.ObjectId },
+  chainToken: { type: mongo.Schema.Types.ObjectId },
+});
+
+export const Stock = mongo.createModel<Types.StockDocument>('Stocks', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: MarketStock

File: modules/market/market.models.ts [hunk 5]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  marketCap: { type: Number },
+  volume: { type: Number },
+  currency: { type: String, required: true },
+  AUM: { type: Number },
+  quoteType: { type: String },
+  dailyChange: { type: Number },
+  dailyChangePercent: { type: Number },
+});
+
+export const ChainToken = mongo.createModel<Types.ChainDocument>('ChainToken', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: MarketToken (its different than a chain token)

File: modules/market/market.models.ts [hunk 6]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  name: { type: String, required: true },
+  symbol: { type: String, required: true },
+  currentPrice: { type: Number, required: true },
+});
+
+export const StockCompany = mongo.createModel<Types.StockCompanyDocument>('StockCompany', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: MarketCompany

File: modules/market/market.models.ts [hunk 7]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  name: { type: String, required: true },
+  ticker: { type: String, required: true },
+  country: { type: String, required: true },
+  industry: { type: String, required: true },
+  sector: { type: String, required: true },
+  type: { type: String, required: true },
+  issuedETFs: [{ type: mongo.Schema.Types.ObjectId, ref: 'ETF' }],
+});
+
+export const ETF = mongo.createModel<Types.ETFDocument>('ETF', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: MarketETF

File: modules/market/market.models.ts [hunk 8]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  name: { type: String, required: true },
+  ticker: { type: String, required: true },
+  issuer: { type: mongo.Schema.Types.ObjectId, required: true },
+  leverage: { type: Number, required: false },
+});
+
+export const StockSentiment = mongo.createModel<Types.StockSentimentDocument>('StockSentiment', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: MarketStockSentiment

File: modules/market/market.models.ts [hunk 9]
```diff
@@ -8,3 +8,113 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  ticker: { type: String, required: true },
+  sentiment: {
+    label: { type: String, required: true },
+    score: { type: Number, required: true },
+  },
+  context: { type: String, required: true },
+  confidence: { type: Number, required: true },
+});
+
+export const BigData = mongo.createModel<Types.BigDataDocument>('BigData', {
```
Comment (hashwarp) [inline modules/market/market.models.ts]: use the Data model from Core?


// Data Model
export const Data = mongo.createModel<Types.DataDocument>(
  'Data',
  {
    mod: { type: String, required: true },
  },
  {
    virtuals: [...addTagVirtuals('Data'), ...addApplicationVirtual()],
  }
);

File: modules/market/market.schema.ts [hunk 10]
```diff
@@ -17,3 +17,137 @@ export const MarketExchange = Entity.merge(
     value: z.string().optional(),
   })
 );
+
+export const MarketAnalysis = Entity.merge(
+  z.object({
+    action: z.enum(['buy', 'sell']), // buy or sell action
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: Buy or Sell

File: modules/market/market.schema.ts [hunk 11]
```diff
@@ -17,3 +17,137 @@ export const MarketExchange = Entity.merge(
...
+  z.object({
+    symbol: z.string(),
+    tokenName: z.string(),
+    currentPrice: z.number(),
+  })
+);
+
+export const StockCompany = Entity.merge(
+  z.object({
+    name: z.string(), // Company name, e.g., "Direxion"
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: Entity has `name` already

File: modules/market/market.schema.ts [hunk 12]
```diff
@@ -17,3 +17,137 @@ export const MarketExchange = Entity.merge(
...
+    industry: z.string().optional(), // Industry the company operates in
+    sector: z.string().optional(), // Sector the company belongs to
+    issuedETFs: z.array(ObjectId).optional(), // References to ETFs issued by this company
+  })
+);
+
+export const ETF = Entity.merge(
+  z.object({
+    ticker: z.string(),
+    name: z.string(), // Name of the ETF, e.g., "Direxion Daily Semiconductor Bull 3X Shares"
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: Entity has `name` already

File: modules/market/market.models.ts (hunk 9 above)
Comment (hashwarp) [inline modules/market/market.models.ts]: already exists, core.models.ts => Data

File: modules/market/market.models.ts (hunk 5 above)
Comment (hashwarp) [inline modules/market/market.models.ts]: already exists, chain.models.ts => ChainToken

File: modules/market/market.schema.ts [hunk 13]
```diff
@@ -17,3 +17,127 @@ export const MarketExchange = Entity.merge(
...
+    sector: z.string().optional(), // Sector the ETF is focused on
+  })
+);
+
+export const MarketStockSentiment = Entity.merge(
+  z.object({
+    ticker: z.string(),
+    company: z.string(),
+    sentiment: z.object({
+      label: z.enum(['positive', 'neutral', 'negative']), // predefined sentiment labels
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: Positive, Neautral, Negative

no lower case, remember?

File: modules/market/market.schema.ts [hunk 14]
```diff
@@ -17,3 +17,127 @@ export const MarketExchange = Entity.merge(
...
+      .array(
+        z.object({
+          date: z.string(), // Date of transaction
+          type: z.enum(['Buy', 'Sell']),
+          price: z.number(), // Price at transaction
+          quantity: z.number(), // Quantity bought or sold
+        })
+      )
+      .optional(), // Optional array for detailed transaction history
+    stock: ObjectId.optional(), // Reference to Stock if type is 'Stock'
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: should be stockId

File: modules/market/market.schema.ts [hunk 15]
```diff
@@ -17,3 +17,127 @@ export const MarketExchange = Entity.merge(
...
+    chainToken: ObjectId.optional(), // Reference to ChainToken if type is 'ChainToken'
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: should be chainTokenId

File: modules/market/market.schema.ts [hunk 16]
```diff
@@ -17,3 +17,127 @@ export const MarketExchange = Entity.merge(
...
+  z.object({
+    categoryGoals: z.array(
+      z.object({
+        category: z.string(), // Category name
+        goalPercentage: z.number(), // Goal percentage for this category
+        currentPnL: z.number().optional(), // Current P&L for this category
+        historicalPnL: z.number().optional(), // Historical P&L for this category
+      })
+    ), // Array of categories with goal percentages and P&L tracking
+    investments: z.array(ObjectId), // Array of ObjectId references to Investment
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: investmentIds

File: modules/market/market.models.ts [hunk 17]
```diff
@@ -8,3 +8,112 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+  labels: {
+    sector: { type: String },
+    topic: { type: String },
+    sentiment: { type: String },
+  },
+});
+
+export const MarketInvestor = mongo.createModel<Types.MarketInvestorDocument>('MarketInvestor', {
+  name: { type: String, required: true },
+  portfolios: { type: mongo.Schema.Types.ObjectId, required: true },
```
Comment (hashwarp) [inline modules/market/market.models.ts]: rename to `portfolioIds`

`portfolios` would be a relationship

File: modules/market/market.models.ts [hunk 18]
```diff
@@ -8,3 +8,112 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+    name: { type: String, required: true },
+    categoryGoals: [
+      {
+        category: { type: String, required: true },
+        goalPercentage: { type: Number, required: true },
+        currentPnL: { type: Number },
+        historicalPnL: { type: Number },
+      },
+    ],
+    investments: { type: mongo.Schema.Types.ObjectId, required: true },
```
Comment (hashwarp) [inline modules/market/market.models.ts]: same here

File: modules/market/market.models.ts [hunk 19]
```diff
@@ -8,3 +8,112 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
...
+
+export const MarketStock = mongo.createModel<Types.MarketStockDocument>('MarketStock', {
+  ticker: { type: String, required: true },
+  companyName: { type: String, required: true },
+  companyId: { type: mongo.Schema.Types.ObjectId, required: true, ref: 'StockCompany' },
+  currentPrice: { type: Number, required: true },
+  marketCap: { type: Number },
+  volume: { type: Number },
+  currency: { type: String, required: true },
+  AUM: { type: Number },
```
Comment (hashwarp) [inline modules/market/market.models.ts]: lower case

File: modules/market/market.schema.ts [hunk 20]
```diff
@@ -17,3 +17,127 @@ export const MarketExchange = Entity.merge(
...
+      .optional(), // Optional array for detailed transaction history
+    stock: ObjectId.optional(), // Reference to Stock if type is 'Stock'
+    chainToken: ObjectId.optional(), // Reference to ChainToken if type is 'ChainToken'
+  })
+);
+
+export const MarketStock = Entity.merge(
+  z.object({
+    ticker: z.string(),
+    companyName: z.string(),
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: dont need company name here, because it should be company.name

File: modules/market/market.schema.ts [hunk 21]
```diff
@@ -17,3 +17,127 @@ export const MarketExchange = Entity.merge(
...
+
+export const MarketStockSentiment = Entity.merge(
+  z.object({
+    ticker: z.string(),
+    company: z.string(),
+    sentiment: z.object({
+      label: z.enum(['positive', 'neutral', 'negative']), // predefined sentiment labels
+      score: z.number().min(0).max(1), // sentiment score as a confidence percentage
+    }),
+    context: z.string(), // description of the sentiment's context
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: use `description`

File: modules/market/market.schema.ts [hunk 22]
```diff
@@ -17,3 +17,127 @@ export const MarketExchange = Entity.merge(
...
+    AUM: z.number().optional(), // Assets under management
+    industry: z.string().optional(), // Industry the ETF is focused on
+    sector: z.string().optional(), // Sector the ETF is focused on
+  })
+);
+
+export const MarketStockSentiment = Entity.merge(
+  z.object({
+    ticker: z.string(),
+    company: z.string(),
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: should be companyId ?

File: modules/market/market.models.ts [hunk 23]
```diff
@@ -8,3 +8,112 @@ export const Market = mongo.createModel<Types.MarketDocument>('Market', {
 export const MarketPair = mongo.createModel<Types.MarketPairDocument>('MarketPair', {
   value: { type: String, required: true },
 });
+
+export const MarketExchange = mongo.createModel<Types.MarketExchangeDocument>('MarketExchange', {
+  value: { type: String },
```
Comment (hashwarp) [inline modules/market/market.models.ts:13]: not needed? delete from others if not needed

Comment (horlami228) [reply modules/market/market.models.ts:13]: i found this one there already

i should remove the two model?

File: modules/market/market.models.ts [hunk 24]
```diff
@@ -1,10 +1,107 @@
...
+    topic: { type: String },
+    sentiment: { type: String },
+  },
 });
 
-export const MarketPair = mongo.createModel<Types.MarketPairDocument>('MarketPair', {
-  value: { type: String, required: true },
+export const MarketInvestor = mongo.createModel<Types.MarketInvestorDocument>('MarketInvestor', {
+  name: { type: String, required: true },
+  portfolioIds: { type: mongo.Schema.Types.ObjectId, required: true },
```
Comment (hashwarp) [inline modules/market/market.models.ts]: plural? is this suppose to be an array?

File: modules/market/market.models.ts [hunk 25]
```diff
@@ -1,10 +1,107 @@
...
+    name: { type: String, required: true },
+    categoryGoals: [
+      {
+        category: { type: String, required: true },
+        goalPercentage: { type: Number, required: true },
+        currentPnL: { type: Number },
+        historicalPnL: { type: Number },
+      },
+    ],
+    investmentIds: { type: mongo.Schema.Types.ObjectId, required: true },
```
Comment (hashwarp) [inline modules/market/market.models.ts]: array?

File: modules/market/market.schema.ts [hunk 26]
```diff
@@ -1,19 +1,123 @@
...
+
+export const MarketStockSentiment = Entity.merge(
+  z.object({
+    ticker: z.string(),
+    companyId: ObjectId,
+    sentiment: z.object({
+      label: z.enum(['Positive', 'Neutral', 'Negative']), // predefined sentiment labels
+      score: z.number().min(0).max(1), // sentiment score as a confidence percentage
+    }),
+    description: z.string(), // description of the sentiment's context
```
Comment (hashwarp) [inline modules/market/market.schema.ts]: description exists on Entity, and this extends Entity

File: modules/market/market.types.ts [hunk 27]
```diff
@@ -11,18 +11,36 @@ export type * from './market.router';
...
-  MarketExchange: Model<MarketExchangeDocument>;
+  MarketAnalysis: Model<MarketAnalysisDocument>;
+  MarketInvestor: Model<MarketInvestorDocument>;
+  InvestmentPortfolio: Model<MarketInvestmentPortfolioDocument>;
+  Investment: Model<MarketInvestmentDocument>;
+  Stock: Model<MarketStockDocument>;
+  ChainToken: Model<MarketTokenDocument>;
+  Company: Model<MarketCompanyDocument>;
+  ETF: Model<MarketETFDocument>;
+  StockSentiment: Model<MarketStockSentimentDocument>;
```
Comment (hashwarp) [inline modules/market/market.types.ts]: ^^^^^^^

File: modules/market/market.models.ts (hunk 24 above)
Comment (horlami228) [reply modules/market/market.models.ts]: i was thinking an investor could potentially have different portfolios but i fixed this in the next commit making this an array same goes for investmentIds

a porfolio could have different investments

**Review by hashwarp — APPROVED** [review]

**Pull request merged**